### PR TITLE
Catch ArrayIndexOutOfBoundsException

### PR DESCRIPTION
**Pull request by snalli**
Catch ArrayIndexOutOfBoundsException when parsing request path

### CODE REVIEW

**Review by ankagrawal — CHANGES_REQUESTED** [review]
The code Chan free looks good. Please add unit test and check if this can be quickly added to an existing integration test (to confirm that in the user request path, everything will be able to pass on the correct exception to the user.

File: ambry-api/src/main/java/com/github/ambry/rest/RequestPath.java [hunk 1]
```diff
@@ -130,6 +130,8 @@ public static RequestPath parse(String path, Map<String, Object> args, List<Stri
           throw new RestServiceException("Segment index given is not an integer", RestServiceErrorCode.BadRequest);
         }
         //otherwise, nothing to do
+      } catch (ArrayIndexOutOfBoundsException e) {
+        throw new RestServiceException("Bad request path", RestServiceErrorCode.BadRequest);
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/rest/RequestPath.java:134]: let's not do this, let's find the code that use out of bound index and check the length of the array,  throw a RestServiceException.

Comment (snalli) [reply ambry-api/src/main/java/com/github/ambry/rest/RequestPath.java:134]: The code that used out-of-bound index is L120
```
if (SEGMENT.equals(fields[fields.length - 2])) {
```

What's the difference between the approach in this PR and the one you are suggesting ? The end result in either case is a RestServiceException.

**Comment by snalli** [comment]
> The code Chan free looks good. Please add unit test and check if this can be quickly added to an existing integration test (to confirm that in the user request path, everything will be able to pass on the correct exception to the user.

added test

File: ambry-api/src/main/java/com/github/ambry/rest/RequestPath.java (hunk 1 above)
Comment (panchow) [reply ambry-api/src/main/java/com/github/ambry/rest/RequestPath.java:134]: Let's lost the root cause exception.  Normally, we if a wrap an other exception, we capture it as the cause.  I notice the other new RestServiceException() doesn't capture the cause exception.  It's a good practice to capture the cause exception.

**Review by panchow — APPROVED** [review]

File: ambry-api/src/main/java/com/github/ambry/rest/RequestPath.java (hunk 1 above)
Comment (snalli) [reply ambry-api/src/main/java/com/github/ambry/rest/RequestPath.java]: good point !

Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/rest/RequestPath.java]: `ArrayIndexOutOfBoundsException` is a RuntimeException. It's a bit like `NullPointerException`. If you already know which logic would yield this exception, then we rely on if-else statement, not try-catch statement. It's faster with if-else.

**Review by ankagrawal — APPROVED** [review]

**Pull request merged**